### PR TITLE
Add `maxTtl` option

### PR DESCRIPTION
**Pull request by szmarczak**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -135,6 +135,13 @@ If set to `false`, after a cached resource's TTL expires it is kept in the cache
 
 If set to `true` once a cached resource has expired it is deleted and will have to be re-requested.
 
+###### opts.maxTtl
+
+Type: `number`<br>
+Default: `Infinity`
+
+Limits TTL.
```
Comment (sindresorhus) [inline README.md]: The description should be a bit more verbose. For example, is the `number` in milliseconds?

Comment (szmarczak) [reply README.md]: yup

File: test/cache.js [hunk 2]
```diff
@@ -540,6 +540,23 @@ test('checks status codes when comparing cache & response', async t => {
 	t.is(secondResponse.body, 'ok');
 });
 
+test('ability to limit TTL', async t => {
+	const endpoint = '/cache';
+	const cache = new Map();
```
Comment (szmarczak) [inline test/cache.js]: Need to change `Map` to something different supporting TTL.

File: src/index.js [hunk 3]
```diff
@@ -97,7 +97,14 @@ class CacheableRequest {
 									statusCode: response.fromCache ? revalidate.statusCode : response.statusCode,
 									body
 								};
-								const ttl = opts.strictTtl ? response.cachePolicy.timeToLive() : undefined;
+								let ttl = opts.strictTtl ? response.cachePolicy.timeToLive() : undefined;
+								if (opts.maxTtl) {
+									if (ttl) {
+										ttl = Math.min(ttl, opts.maxTtl);
+									} else {
+										ttl = opts.maxTtl;
```
Comment (sindresorhus) [inline src/index.js]: I would use a ternary here:

```js
ttl = ttl ? Math.min(ttl, opts.maxTtl) : opts.maxTtl;
```

File: test/cache.js [hunk 4]
```diff
@@ -540,6 +540,32 @@ test('checks status codes when comparing cache & response', async t => {
 	t.is(secondResponse.body, 'ok');
 });
 
+test('ability to limit TTL', async t => {
+	const endpoint = '/cache';
+	const store = new Map();
+	const cache = {
+		get: store.get.bind(store),
+		set: (key, val, ttl) => {
+			t.true(typeof ttl === 'number');
```
Comment (sindresorhus) [inline test/cache.js]:
```js
t.is(typeof ttl, 'number');
```

File: test/cache.js [hunk 5]
```diff
@@ -540,6 +540,32 @@ test('checks status codes when comparing cache & response', async t => {
 	t.is(secondResponse.body, 'ok');
 });
 
+test('ability to limit TTL', async t => {
+	const endpoint = '/cache';
+	const store = new Map();
+	const cache = {
+		get: store.get.bind(store),
+		set: (key, val, ttl) => {
```
Comment (sindresorhus) [inline test/cache.js]: `val` => `value`

File: test/cache.js [hunk 6]
```diff
@@ -350,10 +350,10 @@ test('TTL is passed to cache', async t => {
 	const store = new Map();
 	const cache = {
 		get: store.get.bind(store),
-		set: (key, val, ttl) => {
```
Comment (lukechilds) [inline test/cache.js]: I don't mind you changing this `val` => `value` but can you keep it consistent with all other references in the file?

**Comment by lukechilds** [comment]
Thanks for this @szmarczak, great work, sorry again for the delay in review, personal life has been a bit crazy, really struggling to fit OSS in at the moment.

Left a few comments. Let me know what you think and I'll try and get this merged ASAP for you.

File: test/cache.js (hunk 6 above)
Comment (szmarczak) [reply test/cache.js]: Sure

File: src/index.js [hunk 7]
```diff
@@ -97,7 +97,10 @@ class CacheableRequest {
 									statusCode: response.fromCache ? revalidate.statusCode : response.statusCode,
 									body
 								};
-								const ttl = opts.strictTtl ? response.cachePolicy.timeToLive() : undefined;
+								let ttl = opts.strictTtl ? response.cachePolicy.timeToLive() : undefined;
+								if (opts.maxTtl) {
+									ttl = ttl ? Math.min(ttl, opts.maxTtl) : opts.maxTtl;
```
[the remark this answers is not in the document]
Comment (szmarczak) [reply src/index.js]: I'll look at it

File: test/cache.js (hunk 6 above)
Comment (szmarczak) [reply test/cache.js]: done

File: src/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (szmarczak) [reply src/index.js]: fixed

**Comment by szmarczak** [comment]
ping

**Comment by szmarczak** [comment]
ping x2

**Comment by lukechilds** [comment]
Apologies for the delay in review, I'll go over this in the morning, thanks for your work on this.

**Comment by lukechilds** [comment]
Thanks for this @szmarczak sorry to have messed you around, but regarding my previous comments:

>I think `opts.strictTtl` should be prioritised over `opts.maxTtl`.

I think I had a bit of a brain fart, it should actually be that `opts.maxTtl` should be prioritised over `opts.strictTtl`, **unless** `opts.strictTtl` is true and the remote TTL is smaller than `opts.maxTtl`, then the remote TTL value should be used instead. So your original implementation was actually spot on.

I've reverted the change and added tests to check every combination of:
- `opts.maxTtl`
- `opts.strictTtl` true + `opts.maxTtl` smaller than remote
- `opts.strictTtl` true + `opts.maxTtl` larger than remote

to make sure this behaviour stays the same.

---

I'm happy for this to be merged as it is now, but would be good to get your feedback on whether you agree with what I've said here. If you agree let me know and I'll hit merge and get a new release out.

**Comment by szmarczak** [comment]
No worries! I completely agree ;) Thanks for improving the tests. I think we're ready for the merge too.

**Comment by lukechilds** [comment]
@szmarczak Great, published as `cacheable-request@5.1.0`.

**Pull request merged**